third-party/realdds/src/dds-device-server.cpp

```diff
@@ -4,264 +4,91 @@
-    {
-        // Send all initialization notifications
-        for( auto notification : _init_notifications )
-        {
-            DDS_API_CALL( _data_writer->write( &notification ) );
-        }
-    }
-
-    std::shared_ptr< dds_participant > _participant;
-    eprosima::fastdds::dds::Publisher * _publisher;
-    eprosima::fastdds::dds::Topic * _topic;
-    eprosima::fastdds::dds::DataWriter * _data_writer;
-    dds_notifications_client_listener _clients_listener;
-    active_object<> _notifications_loop;
-    single_consumer_queue< topics::raw::device::notification > _instant_notifications;
-    std::vector< topics::raw::device::notification > _init_notifications;
-    std::mutex _notification_send_mutex;
-    std::condition_variable _send_notification_cv;
-    std::atomic_bool _send_init_msgs = { false }; 
-    std::atomic_bool _new_instant_notification = { false };  
-    std::atomic_bool _active = { false };
-};
-
-//-----------------------------------------------------------------------------------------------------//
-// DDS device server is in charge of handling the device control, streams and notifications
-//-----------------------------------------------------------------------------------------------------//
 dds_device_server::dds_device_server( std::shared_ptr< dds_participant > const & participant,
                                       const std::string & topic_root )
     : _participant( participant )
-    , _publisher( nullptr )
+    , _publisher( std::make_shared< dds_publisher >( participant ))
     , _topic_root( topic_root )
 {
     LOG_DEBUG( "DDS device server for device topic root: " << _topic_root << " created" );
 }
 
+
 dds_device_server::~dds_device_server()
 {
-    // Release resources
-    if( nullptr != _participant )
-    {
-        if( nullptr != _publisher )
-        {
-            _participant->get()->delete_publisher( _publisher );
-        }
-    }
-
     _stream_name_to_server.clear();
-    LOG_DEBUG( "DDS device server for device topic root: " << _topic_root << " deleted" );
+    LOG_DEBUG( "DDS device server '" << _topic_root << "' deleted" );
 }
 
-void dds_device_server::set_image_header( const std::string & stream_name,
-                                          const image_header & header )
+
+void dds_device_server::start_streaming( const std::string & stream_name,
+                                         const image_header & header )
 {
-    _stream_name_to_server.at( stream_name )->set_image_header( header );
+    _stream_name_to_server.at( stream_name )->start_streaming( header );
 }
 
+
 void dds_device_server::publish_image( const std::string & stream_name, const uint8_t * data, size_t size )
 {
-    if( !is_valid() )
-    {
-        throw std::runtime_error( "Cannot publish '" + stream_name + "' frame for '" + _topic_root
-                                  + "', DDS device server in uninitialized" );
-    }
+    auto stream = _stream_name_to_server.at( stream_name );  // may throw!
```

Comment:
Isn't it better to catch or `find` and use an explicit `DDS_THROW` message?

Reply:
Probably... but this is a leftover from previous code.
I changed to:
```
    auto it = _stream_name_to_server.find( stream_name );
    if( it == _stream_name_to_server.end() )
        DDS_THROW( runtime_error, "stream '" + stream_name + "' does not exist" );
    auto & stream = it->second;
```

Reply:
(in next PR)